[lgwebos] Initial contribution of LG webOS binding. Closes #2305

sprehn opened the pull request:
Code Review for this new binding was already in progress. In order to fix old commit messages without Sign-off I reset my fork of openhab2-addons and reintegrated my binding code. The reset automatically closed my earlier PR. Please continue the review.

sprehn commented:
@Hilbrand,  @ppieczul could you please continue your review here

ppieczul commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,221 @@
+
+    @Override
+    public String getApplicationName() {
+        return "Openhab";
```
Unless this forces a convention of first capital letter, please use openHAB wording.

sprehn replied:
np

ppieczul commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,221 @@
+
+    @Override
+    public String getPackageName() {
+        return "org.openhab";
```
Can you use getClass().getPackage().getName() here? Just to avoid hard coded values.

sprehn replied:
this would return org.openhab.binding.lgwebos.internal.discovery. In combination with ApplicationName this is used by LG TVs to identify the client application.
We don't want the actual package name. We need a constant namespace here. That's why I think org.openhab is better, as it would not change when we refactor LGWebOSDiscovery. 

But I'll add a comment to the code, why.

ppieczul commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,221 @@
+                logger.debug("localIP parameter explicitly set to: {}", localIP);
+                return InetAddress.getByName(localIP.trim());
+            } catch (UnknownHostException e) {
+                logger.warn("localIP config parameter could not be parsed: {}", localIP);
```
If you show exception's getMessage(), there will be detailed explanation what's wrong with the localIP value.

sprehn replied:
will add

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<description>WebOS based smart TV</description>
+
+		<channels>
+			<channel id="power" typeId="commandType">
```
yes, TV powers off when items changes from ON to OFF. When the TV turns on the binding will detect it a few seconds later. It will then update this channel and set it to ON.

ppieczul replied:
My question is more about when the switch is in OFF state and I push it in the UI. It will send ON command and show it is in ON state until a state update is received from the binding. If ON command will actually not turn the TV on, the real TV state will be OFF, but the UI will show it as ON. I don't know if the state will be updated from the TV. If not, it may need some code in the power control to handle this.

sprehn replied:
whenever the tv turns on or off the channel will be updated correctly.

however, if somebody switches to ON in the UI the switch will go to ON, but nothing happens on the tv. This Is not supported LG TVs (when they are off, there is no api). On some models, like mine it can be solved with the Wake On LAN binding as suggested in the README documentation.

ppieczul replied:
Would it work if for the ON command, we simply update the channel state to the current TV state (which should be OFF really, but someone may start the TV from the outside in the meantime)? This should cause the switch to return to OFF.

sprehn replied:
I added autoupdate="false" to the demo item. that should do the trick.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/.classpath:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Hm. I am using the recommended IDE and ran an update to be sure. I went through all xml files and executed the auto formatter. It applied some changes, but it did not replace tabs with spaces. I also checked a couple other .classpath files. They also use tabs.

martinvw replied:
Indeed tabs are the way to go for XML files in our projects.

ppieczul replied:
Ok, so it is my IDE that incorrectly replaces tabs with spaces in xmls. Sorry for confusion.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+			<channel id="channelName" typeId="channelNameChannelType" />
+			<channel id="toast" typeId="toastChannelType" />
+			<channel id="mediaPlayer" typeId="mediaPlayerType" />
+			<channel id="mediaStop" typeId="commandType">
```
I only set a common channel type for those channels which used On/Off to send a specific command. The channels still have their label and descriptions in the type definition. 

mediaStop has a label but no description. Are you suggesting I should provide a description ?

ppieczul replied:
What I meant is that for all channels which have commandType type, there is a label and description provided. But not for mediaStop.

sprehn replied:
I think it is the incorrect indention that is confusing here. formatting is fixed now and I also added a little description.

ppieczul replied:
You are right! I looked at that several times and just have not absorbed the missing / in the closing >.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+		<label>Channel Number</label>
+		<description>Current Channel Setting</description>
+	</channel-type>
+	<channel-type id="toastChannelType">
```
what would you suggest as a name for a common type? the commonality is that both are of type string.

ppieczul replied:
How about stringType

sprehn replied:
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,82 @@
+	   <label>Media Control</label>
+	   <description>Control media (e.g. audio or video) playback</description>
+	</channel-type>
+	<channel-type id="appLauncherChannelType">
```
true

sprehn replied:
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,190 @@
+| --- | --- |
+| LocalIP |  This is the local IP of your OpenHAB host on the network. (Optional) |
+
+The binding will attempt to auto detect your IP, if LocalIP is not set. This works when your hostname resolves to this IP and not to the loopback interface or, if the system has exactly one non-loop back network interface. Otherwise this has to be explicitly set. If you are unable to discover devices please check the log file for error messages.e.g.: 
```
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,190 @@
+
+## Discovery
+
+TVs are auto discovered through SSDP in the local network. The binding broadcast a search message via UDP on the network. 
```
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,190 @@
+| --------------- | ------------ | ------------ | ---------- |
+| power | Switch | Current power setting. TV can only be powered off, not on.  | RW |
+| mute | Switch | Current mute setting.  |  RW |
+| volume | Dimmer | Current volume setting. Setting and reporting absolute percent values only works when using internal speakers. Connected to an external amp the volume should be controlled using increase and decrease relative commands. |  RW |
```
agreed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
````diff
@@ -0,0 +1,190 @@
+demo.items:
+
+```
+Switch LG_TV0_Power "TV Power" <television> { channel="lgwebos:WebOSTV:192_168_2_119:power" }
````
It is actually not required. Removed it.

ppieczul replied:
I though this was to simulate a pushbutton, not on/off switch. This would make sense, as it looks like these controls send commands to the TV on any channel command received. When using a UI like habPanel, this is not a problem, as we can program easily a pushbutton that sends only ON commands. But for other UIs there may be a switch displayed which will toggle between ON and OFF and on each toggle send command to the TV. Adding autoupdate=false will cause the switch to immediately return to the original state after it is pushed to ON, unless binding explicitly updates its state, which will never happen. So it would behave like a pushbutton, which seems to be correct here.

sprehn replied:
autoupdate="false" added

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    @ActionDoc(text = "opens the application with given appId and passes additional parameters")
```
I have added a paragraph on launchApplicationWithParams. The others are very experimental (hidden) features. If that is a problem I would rather remove the code from the binding.

sprehn replied:
And it should be launchApplicationWithParam, not launchApplicationWithParams, i guess.

ppieczul replied:
I see no problem with these features in the code, I just thought these were overlooked in the readme.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+        if (control == null) {
+            return Collections.emptyList();
+        }
+        BlockingQueue<List<String>> result = new ArrayBlockingQueue<>(1);
```
the results get populated in an asynchronous callback (onError or onSuccess). but in this case I need to block until I have the values back.

sprehn replied:
method removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+                try {
+                    result.put(Collections.emptyList());
+                } catch (InterruptedException e) {
+                    LOGGER.warn("interruppted", e);
```
sure

sprehn replied:
method removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    @ActionDoc(text = "sends the ender key to a web os device")
```
omg - a little embarrassed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+/**
```
I have to admit I have not fully understood the reason behind the packet structure and difference between package and package.internal. Now that you mention the exports this I realize it. I agree.
BaseChannelHandler and all its implementations don't have to be exported and could go to internal. Will move them now.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("internalReceiveCommand({},{}) is called", channelUID, command);
```
true

sprehn commented:
ok, will add autoupdate=false to the demo items
> Am 07.10.2017 um 21:13 schrieb Pawel Pieczul <notifications@github.com>:
> 
> @ppieczul commented on this pull request.
> 
> In addons/binding/org.openhab.binding.lgwebos/README.md <https://github.com/openhab/openhab2-addons/pull/2768#discussion_r143338360>:
> 
> > +```
> +
> +Opens the application with given appId. To find out what appId constant matches which app, bind the appLauncher channel to a String item and turn the TV to the desired application.
> +
> +The first parameter is the IP address of your TV. 
> +The second parameter is the application id that you want to open.
> +
> +
> +## Full Example
> +
> +This example assumes the IP of your smart TV is 192.168.2.119.
> +
> +demo.items:
> +
> +```
> +Switch LG_TV0_Power "TV Power" <television> { channel="lgwebos:WebOSTV:192_168_2_119:power" }
> I though this was to simulate a pushbutton, not on/off switch. This would make sense, as it looks like these controls send commands to the TV on any channel command received. When using a UI like habPanel, this is not a problem, as we can program easily a pushbutton that sends only ON commands. But for other UIs there may be a switch displayed which will toggle between ON and OFF and on each toggle send command to the TV. Adding autoupdate=false will cause the switch to immediately return to the original state after it is pushed to ON, unless binding explicitly updates its state, which will never happen. So it would behave like a pushbutton, which seems to be correct here.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openhab/openhab2-addons/pull/2768#discussion_r143338360>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ACafwIohcdRwIWkN64ArG_YeQ1VGYHL9ks5sp81qgaJpZM4Pw_8D>.
> 

—
Sebastian Prehn
Huswertstraße 30
60435 Frankfurt am Main

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+            device.removeListener(this);
+        }
+        this.discoveryManager.removeListener(this);
+        this.discoveryManager = null;
```
So I'll remove the this.discoveryManager = null;

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+    /**
+     * Refresh channel subscription for one specific channel.
+     *
+     * @param channelUID must not be <code>null</code>
```
Where can I read about this? not familiar with `@NonNullByDefault`

ppieczul replied:
The has recently been added a [guideline](https://github.com/triller-telekom/smarthome/blob/e4cb58328c62c74b1f4df9fca17bf9f8f4bd8110/docs/documentation/development/guidelines.md) to the smarthome framework - point A.12.
OH bindings can (and maybe will have to, soon) follow these guidelines. IDE will then detect incorrect, redundant and missing handling of nulls. There is however an issue with calls to the libraries, which are not aware of the annotations, like jdk. These places can have warning suppressed. Would be interesting to see if enabling this on your classes shows benefits or is this inconvenience.

sprehn replied:
I read the guideline. Could you recommend an example binding to look at?

ppieczul replied:
There seem to be no bindings that currently implement that. These annotations are gradually introduced to the smarthome framework, so you may just search for NonNullByDefault in your workspace and find examples. Another example, though not finished yet, is my attempt to bring them into my binding: #2604. 
There is also an ongoing [discussion about observations in the smarthome project](https://github.com/eclipse/smarthome/issues/4321).

I think it may make no point to delay merge of this PR and wait for the null annotations, but I would like to welcome you to the discussions and share your observations.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/MediaControlPlayer.java:
```diff
@@ -0,0 +1,70 @@
+        } else if (command instanceof PlayPauseType) {
+            if (PlayPauseType.PLAY.equals(command) && d.hasCapabilities(MediaControl.Play)) {
+                getMediaControl(d).play(createDefaultResponseListener());
+            }
```
can do

ppieczul commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+ */
+@Component(service = ActionService.class, immediate = true, name = "action.lgwebos")
+public class LGWebOSAction implements ActionService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LGWebOSAction.class);
```
I thought about this for a while and it also says 'wherever possible'. Also following the link from that page it says:
`We used to recommend that loggers members be declared as instance variables instead of static. After further analysis, we no longer recommend one approach over the other.`

Maybe as long as methods are static, it makes sense for the logger to be static. There obviously are lots of static loggers all over the current code. What do you think?

sprehn replied:
I know. found this also during the static code analysis. But to log form static action methods what else can I do?

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/MediaControlPlayer.java:
```diff
@@ -0,0 +1,70 @@
+        } else if (command instanceof RewindFastforwardType) {
+            if (RewindFastforwardType.FASTFORWARD.equals(command) && d.hasCapabilities(MediaControl.FastForward)) {
+                getMediaControl(d).fastForward(createDefaultResponseListener());
+            }
```
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/PowerControlPower.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public void onDeviceRemoved(ConnectableDevice device, final String channelId, final LGWebOSHandler handler) {
+        super.onDeviceReady(device, channelId, handler);
```
oh! right

sprehn replied:
this is btw the point where power ON  / OFF gets posted

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/ToastControlToast.java:
```diff
@@ -0,0 +1,55 @@
+            final String value = command.toString();
+            final ToastControl control = getControl(d);
+            try {
+                BufferedImage bi = ImageIO.read(getClass().getResource("/openhab-logo-square.png"));
```
I know, there is some duplication, but I would rather not do right now. the actions handle device ID a little different. and I liked the fact that there is no dependency between binding -> action. The dependency in the other way already exists. But I am not clear how I can get a hold of the toast channel.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,221 @@
+ * @since 1.8.0
+ */
+@Component(service = { DiscoveryService.class, LGWebOSDiscovery.class }, immediate = true, name = "binding.lgwebos")
+public class LGWebOSDiscovery extends AbstractDiscoveryService implements DiscoveryManagerListener, Context {
```
LGWebOSDiscovery is for the action to find it. 
I set the names so that it is 1:1 compared to the xml config file.
Aren't the names important to retrieve the config variable "localIP" correctly in LGWebOSDiscovery (binding.lgwebos)?
I probably don't care about the name of LGWebOSAction and of LGWebOSHandlerFactory

ppieczul replied:
Indeed, I missed that you reference this service. For the config I think you could use configurationPid argument of the component annotation.

sprehn replied:
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,221 @@
+    private InetAddress localInetAddresses;
+
+    public LGWebOSDiscovery() {
+        super(LGWebOSBindingConstants.SUPPORTED_THING_TYPES_UIDS, 60, true);
```
done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+            }
+        });
+        try {
+            return result.take();
```
That's why onError puts an empty list into the queue to be sure that it returns. I am not aware how I could set a timeout.

This is an undocumented / hidden action to scan the TV for all apps on it. I am not recommending it to users.

sprehn replied:
action removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,235 @@
+        });
+        try {
+            return result.take();
+        } catch (InterruptedException e) {
```
ah, I see your point above. if there may be situations where take does not return.

I am tempted to just remove the whole method as it is an easter egg and  I don't know how to make this bullet proof.

sprehn replied:
removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+            if (device.isConnected()) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connected");
+            } else {
+                device.connect();
```
I'll add updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connecting ..."); 

on success onDeviceReady will be called,if pairing is required onPairingRequired, otherwise onConnectionFailed
all set the appropriate status

sprehn replied:
I have also added a comment into the code similar to the above

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,223 @@
+    public void onDeviceReady(ConnectableDevice device) { // this gets called on connection success
+        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connected");
+        refreshAllChannelSubscriptions(device);
+        channelHandlers.entrySet().forEach(e -> e.getValue().onDeviceReady(device, e.getKey(), this));
```
true, done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/VolumeControlMute.java:
```diff
@@ -0,0 +1,72 @@
+            OnOffType onOffType;
+            if (command instanceof OnOffType) {
+                onOffType = (OnOffType) command;
+            } else if (command instanceof StringType) {
```
you are right, removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/MediaControlPlayer.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public void onReceiveCommand(final ConnectableDevice d, Command command) {
+        if (command instanceof NextPreviousType) {
+            if (NextPreviousType.NEXT.equals(command) && d.hasCapabilities(PlaylistControl.Next)) {
```
Done, also for the other handlers

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/ToastControlToast.java:
```diff
@@ -0,0 +1,55 @@
+            final ToastControl control = getControl(d);
+            try {
+                BufferedImage bi = ImageIO.read(getClass().getResource("/openhab-logo-square.png"));
+                ByteArrayOutputStream os = new ByteArrayOutputStream();
```
will do

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/MediaControlPlayer.java:
```diff
@@ -0,0 +1,70 @@
+        if (command instanceof NextPreviousType) {
+            if (NextPreviousType.NEXT.equals(command) && d.hasCapabilities(PlaylistControl.Next)) {
+                getPlayListControl(d).next(createDefaultResponseListener());
+            }
```
yes

ppieczul commented:
Hi @sprehn 
This has been a really fast progress on this PR! I'd like to summarize the couple of remaining items from my point of view: 
  * This poweroff case - I think it is worth to prevent the inconsistency between item state and TV state, so item state always shows actual value.
  * The configurationPid in service arguments instead of name - both will work but this will be consistent with other bindings.
Thank you
Pawel

sprehn commented:
Hi @ppieczul 

both addressed now.

and thank you for your detailed review, learned quite a bit.

Sebastian

ppieczul commented:
Thanks a lot! One thing came to my mind - the localization file, does it need some update after all the changes to the channel types?

sprehn commented:
That's true. Will have a look

sprehn commented:
How can I localize the label and description of a channel?

this worked for channel-type
channel-type.lgwebos.channelNameChannelType.label = Kanal Name
channel-type.lgwebos.channelNameChannelType.description = Kanal Name

but for channel this does not work:
channel.lgwebos.channel.label = Kanal
channel.lgwebos.channel.description = Kanal

ppieczul commented:
I have never tried that, but take a look at [this file](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nUtil.java)

It seems you need to start with `thing-type` add thing type ID to the path after binding ID, since channels are inside thing types.

```
    public String getChannelLabel(Bundle bundle, ThingTypeUID thingTypeUID, ChannelDefinition channel,
            String defaultLabel, Locale locale) {
        String key = I18nUtil.isConstant(defaultLabel) ? I18nUtil.stripConstant(defaultLabel)
                : inferThingTypeKey(thingTypeUID, channel, "label");
        return i18nProvider.getText(bundle, key, defaultLabel, locale);
    }

    public String getChannelDescription(Bundle bundle, ThingTypeUID thingTypeUID, ChannelDefinition channel,
            String defaultDescription, Locale locale) {
        String key = I18nUtil.isConstant(defaultDescription) ? I18nUtil.stripConstant(defaultDescription)
                : inferThingTypeKey(thingTypeUID, channel, "description");
        return i18nProvider.getText(bundle, key, defaultDescription, locale);
    }

    private String inferThingTypeKey(ThingTypeUID thingTypeUID, ChannelDefinition channel, String lastSegment) {
        return "thing-type." + thingTypeUID.getBindingId() + "." + thingTypeUID.getId() + "." + channel.getId() + "."
                + lastSegment;
    }
```

martinvw commented:
@cweitkamp did also a lot of translations recently so maybe he can give you more hints / examples of needed.

cweitkamp commented:
I am afraid that is not possible. I remember having discussed this a while ago but I cannot find the link right now.

In the comments of another PR (https://github.com/openhab/openhab2-addons/pull/2698#discussion_r140665029) you can see that - if I understand it correctly - the usage of labels & descriptions on channels is deprecated. You have to define a separate channel-type for each channel.

ppieczul commented:
I can't argue whether this is the direction framework is going to or not, but I would like to express my opinion that having to define a channel type for each channel separately kind of makes the channel types a redundant entity. Isn't this example really about duplicating the labels and descriptions for channels that already use dedicated channel types?

sprehn commented:
I cannot leave it like this. Half English, half German. 
I can absolutely live without the German localization. Power, Volume, Stop, Channel, ... hey all do not really  require a German translation, and after all it would just be one language.
In order to progress this pull request I'll remove the localization file form this PR.

Hilbrand commented:
The channel and channel-type separation makes sense if you have multiple devices (things) using the same but not all channels. I've a binding with 10 things with similar features. Each thing has a set of channels that point to channel-types. This makes a clear separation. If you only have 1 thing this might seem as overhead. But it makes the overall architecture simpler imho. I would create a separate file for channels and a file for the thing with a one-line-per-channel list that gives a nice clear overview of capabilities.

ppieczul commented:
This makes sense. However, it also allows for pointing multiple channels at the same type, which limits the duplication of data, and today they all can override the label and description. 

The `getChannelLabel` method that takes into account the channel from the thing-type is indeed not used anywhere at all. But when I look at the `createLocalizedThingType` method, which localizes channel definitions, it uses the channel type for the localized data, but as the default it takes the data from the channel definition, not channel type. I have an impression, this should actually use the other method with thing type and channel definition as arguments. I will follow up with that in the smarthome project.
```
            String channelLabel = this.thingTypeI18nUtil.getChannelLabel(bundle, channelDefinition.getChannelTypeUID(),
                    channelDefinition.getLabel(), locale);
            String channelDescription = this.thingTypeI18nUtil.getChannelDescription(bundle,
                    channelDefinition.getChannelTypeUID(), channelDefinition.getDescription(), locale);
```

cweitkamp commented:
@sprehn You can try to [use custom keys](http://www.eclipse.org/smarthome/documentation/features/internationalization.html#using-custom-keys) for translating those channels.

@ppieczul Please keep me up-to-date if you open an issue or contribute in ESH.

sprehn commented:
@cweitkamp thx, but I am absolutely fine without this localization.

cweitkamp commented:
> But when I look at the createLocalizedThingType method, which localizes channel definitions, it uses the channel type for the localized data, but as the default it takes the data from the channel definition, not channel type. I have an impression, this should actually use the other method with thing type and channel definition as arguments.

@ppieczul You are right. Changing those lines like you suggested fixes the problem. I wrote a basic unit test case for the translation of channel definitions: https://github.com/cweitkamp/smarthome/tree/bugfix-channel-i18n. It fails with the current code and succeds with your suggested changes.

ppieczul commented:
@cweitkamp do you plan to bring this into the smarthome project? If it helps you, I could do it too.

cweitkamp commented:
@ppieczu I  contributed over there and added you as additional participant as the suggested changes results from your research.

https://github.com/eclipse/smarthome/pull/4411

sprehn commented:
Hey,

I get warnings when I compile my code now. Here is an example.

`[WARNING] /Users/sprehn/Code/sprehn-openhab2-addons/addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:[69] 
	super(thing);
	      ^^^^^
Null type safety (type annotations): The expression of type 'Thing' needs unchecked conversion to conform to '@NonNull Thing'`

Where can I read more about what I need to implement to resolve those warnings?

ppieczul commented:
Probably [here](https://github.com/eclipse/smarthome/issues/4321) plus google for Java null annotations.

sprehn commented:
I just rebased again to latest upstream and updated my IDE. I don't get any warnings in the IDE. 
Added  org.eclipse.jdt.annotation;resolution:=optional, to the Manifest file, but still I cannot use  @Nullable or @NonNull anywhere.
Just tried to compile other bindings (e.g. avmfritz) and they have even more warnings. So I assume it is not my issue for the moment.

ppieczul commented:
What do you mean by that you can’t use these annotations anywhere?

sprehn commented:
Added org.eclipse.jdt.annotation;resolution:=optional, to the Manifest file.
import org.eclipse.jdt.annotation.NonNull;
Add the @NonNull to a declaration.

Just realized that now it compiles fine without those wornings on commandline, but my Eclipse shows errors. Do I have to reinstall OH Eclipse again?

ppieczul commented:
I don’t know. It happened to me many times that eclipse degraded up to a point reinstallation was needed. In the worst cases reinstallation was not enough - it was also needed to remove the hidden directories in the home dir that eclipse used.

sprehn commented:
Well, just reinstalled in parallel, but now all projects are red. Will call it a day for now. n8

ppieczul commented:
I have the same situation today - fresh and clean install of IDE gives >50K errors. This happens from time to time to me. 
Would be good to hear some comment on that from the veterans - what is causing this (and what is causing this to go away without any particular reason)? It can be really annoying, I'd love to help get rid of it, just don't know even where to start.

sprehn commented:
Yes, my setup (Mac 64bit) is now completely broken. Tried to install the IDE for Mac according to the documentation three times but cannot get it to work.

If I try to start openHAB_Runtime.launch it errors 
An error has occurred.  See the log file
/Users/sprehn/Code/openhab2-master2/ws/.metadata/.plugins/org.eclipse.pde.core/openHAB_Runtime/1507980110410.log

java.lang.ClassNotFoundException: org.eclipse.core.runtime.adaptor.EclipseStarter

All project code is red unable to find the openhab classes.

sprehn commented:
PR #557 fixed the issue with my IDE and I added fixes for those null check warnings.

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/BaseChannelHandler.java:
```diff
@@ -0,0 +1,106 @@
+    // IP to Subscriptions map
+    private Map<String, ServiceSubscription<T>> subscriptions;
+
+    // lazy init
```
Why, I see no benefit?

sprehn replied:
Not all channels support subscriptions. So I figured it would not be good to allocate a hashmap for every channel if it is never used in some cases

sprehn replied:
I'll change return type of getSubscription to Optional to make this clearer.

martinvw replied:
I believe we are not using `Optional` in this project but stick to `null`s and handling them were needed with some help of the annotations.

sprehn replied:
I would like to leave it as is now as it  is standard Java8 and it is actually required to work with optionals when working with the stream and findFirst approach. After all, it is completely binding internal.

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/BaseChannelHandler.java:
```diff
@@ -0,0 +1,106 @@
+    // lazy init
+    private synchronized Map<String, ServiceSubscription<T>> getSubscriptions() {
+        if (subscriptions == null) {
+            subscriptions = new ConcurrentHashMap<String, ServiceSubscription<T>>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

sprehn replied:
Done

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/LauncherApplication.java:
```diff
@@ -0,0 +1,95 @@
+
+                @Override
+                public void onError(ServiceCommandError error) {
+                    logger.warn("error requesting application list: {}.", error.getMessage());
```
Should we mark the device offline?

sprehn replied:
No, errors here could have other reasons, e.g. invalid request syntax. The device gets reliably marked offline already.

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/MediaControlPlayer.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public void onReceiveCommand(final ConnectableDevice d, String channelId, LGWebOSHandler handler, Command command) {
+        if (d == null) {
```
please rename `d` to `device`

martinvw replied:
Appears on many more occurences

sprehn replied:
Done

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/MediaControlPlayer.java:
```diff
@@ -0,0 +1,71 @@
+        if (d == null) {
+            return;
+        }
+        if (command instanceof NextPreviousType) {
```
Are these `instanceof` checks needed, what would you think of something like

```java
switch(command) {
   case NextPreviousType.NEXT:
       if (d.hasCapabilities(PlaylistControl.Next)) {
           getPlayListControl(d).next(createDefaultResponseListener());
      }
      break;

...
```

Or just 6 if statements with a single else which catches the remaining state?

sprehn replied:
Compile error for switch(command). Cannot switch on type Command.
But I can go without the instanceof checks, that's true - Done

martinvw replied:
Yes of course you can only switch on enum types and given that the type of variable is actually an interface which is implemented by different enum's this is not going to work, sorry.

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/PowerControlPower.java:
```diff
@@ -0,0 +1,72 @@
+ * @since 1.8.0
+ */
+public class PowerControlPower extends BaseChannelHandler<Void> {
+    private Logger logger = LoggerFactory.getLogger(PowerControlPower.class);
```
Imho its better to have the logger's final, I hope you will never change them and imho it creates less noise then adding them to parameters.

sprehn replied:
Done 10x

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/TVControlChannel.java:
```diff
@@ -0,0 +1,94 @@
+import com.connectsdk.service.capability.TVControl.ChannelListener;
+import com.connectsdk.service.command.ServiceCommandError;
+import com.connectsdk.service.command.ServiceSubscription;
+import com.google.common.collect.Iterables;
```
Do we really need this (dependency) imho the Java 8 way is more standard and should be prefered can you take a quick look how hard it is to switch to the normal Java 8 of doing such things?

sprehn replied:
Well this stems from earlier versions when java 8 was not the standard in OH yet, but sure.. Multiple occurrences replaced.

sprehn replied:
and now got rid of both google commons dependencies:
```
- com.google.common.base,
- com.google.common.collect,
```

martinvw replied:
👍

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,226 @@
+@Component(service = { DiscoveryService.class,
+        LGWebOSDiscovery.class }, immediate = true, configurationPid = "binding.lgwebos")
+public class LGWebOSDiscovery extends AbstractDiscoveryService implements DiscoveryManagerListener, Context {
+    private static final int DISCOVERY_TIMEOUT = 5; // in seconds
```
what about 

```java
private static final int DISCOVERY_TIMEOUT_SECONDS = 5; // no comment needed ;-)
```

sprehn replied:
Done 👍

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    // Context for connect sdk
+    private final String DATA_DIR = new File("etc" + File.separator + "connect_sdk").getAbsolutePath();
```
1) Put all fields on top
2) where do these files end up? The only place where you should write data is in the userdata, [see for example](https://github.com/openhab/openhab2-addons/search?utf8=%E2%9C%93&q=getUserDataFolder&type=)

sprehn replied:
It was under userdata/etc/connect_sdk. 
But I see the userdata folder can be reconfigured to a different location. So I changed it.

martinvw commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,226 @@
+     */
+    private InetAddress findLocalInetAddresses(String localIP) {
+        // evaluate optional localIP parameter, can be configured through config admin (lgwebos.cfg)
+        if (localIP != null && !localIP.trim().isEmpty()) {
```
If you already have `StringUtils` from `common-lang` available you can just use `StringUtils.isNotBlank()` note that dependency is available in every smarthome install so you can always use it.

sprehn replied:
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/.classpath:
```diff
@@ -0,0 +1,13 @@
+	<classpathentry kind="src" path="src/main/java" />
+	<classpathentry kind="src" path="src/main/resources" />
+	<classpathentry kind="lib" path="lib/java-websocket-1.3.2.jar" />
+	<classpathentry kind="lib" path="lib/json-20140107.jar" />
```
yes I am aware. But this and the two other libs a dependencies of connect sdk.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/.classpath:
```diff
@@ -0,0 +1,13 @@
+	<classpathentry kind="lib" path="lib/java-websocket-1.3.2.jar" />
+	<classpathentry kind="lib" path="lib/json-20140107.jar" />
+	<classpathentry kind="lib" path="lib/xpp3-1.1.4c.jar" />
+	<classpathentry kind="lib" path="lib/Connect-SDK-Java-Core-1.0.jar" />
```
Since I don't own all the devices that this documentation claims are supported I cannot test or troubleshoot it.
I only own an Apple TV and it did not work at all. 
Therefore I renamed the binding from initial connectsdk to lgwebos. a) connectsdk is no longer maintained b) users don't care what the underlying technology is, but what device it supports.

martinvw replied:
Pitty I was hoping for the support for my Apple TV, what version do you have the one with or without apps?

sprehn replied:
it is the older version and when lg released all connectsdk developers the new appletv was not yet released.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+		<description>WebOS based smart TV</description>
+
+		<channels>
+			<channel id="power" typeId="commandType">
```
I used to have one channel type per channel, but I was told in this review that I should combine those command types, as they are all essentially comparable to a button press on a remote control - all modeled in the same way.
Just need a define answer how it should be done.

martinvw replied:
@kaikreuzer can you enlighten us :-)

kaikreuzer replied:
>  I was told in this review that I should combine those command types

Do you have a link to this comment? It is clearly wrong and what @martinvw mentions here is the way it should be done. Sorry, if there were any different requests than this that caused you unneeded efforts...

ppieczul replied:
It is obviously my fault, or incorrect interpretation of the channel type intention and lack of knowledge of this instruction you provided.
I withdraw my comment, however still not entirely understand the reasoning behind why using common types wherever they apply is not correct - what are the wrong implications caused by such approach?

kaikreuzer replied:
@ppieczul I didn't mean to blame anyone, I very much appreciate your help on the reviewing side and it is clear that there is quite some learning curve for all of us until we all have the same understanding of the best modelling!
The issue here is with the understanding what a "common type" should be - a channel type should have a semantic meaning of what kind of function it provides. Reducing it to primitive types such as number/string/boolean would work against this.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,38 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
```
I ran auto-format and this is the change:
```Import-Package: 
  com.google.common.base,
  com.google.common.collect,
- org.eclipse.jdt.annotation;resolution:=optional,
  javax.jmdns,
  org.apache.commons.lang,
  org.apache.http.client.methods,
  org.apache.http.conn.util,
  org.apache.http.message,
  org.apache.http.protocol,
+ org.eclipse.jdt.annotation;resolution:=optional,
  org.eclipse.smarthome.config.core,
  org.eclipse.smarthome.config.discovery,
  org.eclipse.smarthome.core.library.types,```

martinvw replied:
@maggu2810 changed it even further, see https://github.com/openhab/openhab2-addons/pull/2720

maggu2810 replied:
@martinvw Sorry, didn't found time to cleanup the code and place it to a public repo. But I assume as all imports are already ordered alphabetical it should be easy to add new ones on the correct position.
I don't read all the openhab2-addons repository mails I receive, so if you want notify me a `@maggu2810` would not help.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,202 @@
+
+#### TV Settings
+
+The TV must be connected to the same network as OpenHAB with a permanent IP address. If the IP changes the binding will discover it as a different device. 
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,202 @@
+
+The binding has only one configuration parameter, which is only required if the binding cannot automatically detect OpenHAB's local IP address: 
+
+| Name | Description |
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,202 @@
+
+## Channels
+
+| Channel Type ID | Item Type    | Description  | Read/Write |
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/README.md:
```diff
@@ -0,0 +1,202 @@
+Sends a toast message to a webOS device using OpenHAB's logo as an icon.
+The first parameter is the IP address of your TV. 
+The second parameter is the message you want to display.
+### Show Toast with Custom Icon
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/pom.xml:
```diff
@@ -0,0 +1,24 @@
+    <version>2.2.0-SNAPSHOT</version>
+  </parent>
+
+  <properties>
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,191 @@
+/**
+ * This class provides a rules action to start applications and send toast messages.
+ *
+ * @author Sebastian Prehn
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,191 @@
+ * This class provides a rules action to start applications and send toast messages.
+ *
+ * @author Sebastian Prehn
+ * @since 2.1.0
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,191 @@
+public class LGWebOSAction implements ActionService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LGWebOSAction.class);
+
+    private static LGWebOSDiscovery discovery;
```
I think they need to be static so that I can use them in the static methods of an action. Is it not the case for Action methods?

sprehn replied:
I actually tried it again and those methods have to be static.

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,191 @@
+
+    @ActionDoc(text = "sends a toast message to a web os device with openhab icon")
+    public static void showToast(@ParamDoc(name = "deviceId") String deviceId,
+            @ParamDoc(name = "text") final String text) throws IOException {
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/action/lgwebos/internal/LGWebOSAction.java:
```diff
@@ -0,0 +1,191 @@
+            try (ByteArrayOutputStream os = new ByteArrayOutputStream();
+                    OutputStream b64 = Base64.getEncoder().wrap(os);) {
+                ImageIO.write(bi, "png", b64);
+                control.showToast(text, os.toString("UTF-8"), "png", responseListenerObject);
```
Done 2x

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/LGWebOSBindingConstants.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class LGWebOSBindingConstants {
+
+    public static final @NonNull String BINDING_ID = "lgwebos";
```
Otherwise I get a warning on the next line regarding Null safety where this constant is used:
```public static final ThingTypeUID THING_TYPE_WEBOSTV = new ThingTypeUID(BINDING_ID, "WebOSTV");```

martinvw replied:
Given this is so obvious @NonNull already I would personally prefer to leave out the annotation, @triller-telekom wdyt?

triller-telekom replied:
In out [coding guidelines in point 12](https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style) we agreed that IF one chooses to use null annotations he has to use @NonNullByDefault on the class level and only @NonNull on variables/method parameters if they can be null.

Regarding the warning/error in the IDE, we are still consulting the setting which we will use for the null analysis, see our discussion here: https://github.com/eclipse/smarthome/issues/4321

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/LGWebOSBindingConstants.java:
```diff
@@ -0,0 +1,46 @@
+
+    public static final @NonNull String BINDING_ID = "lgwebos";
+
+    // List of all Thing Type UIDs
```
Removed

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/LGWebOSBindingConstants.java:
```diff
@@ -0,0 +1,46 @@
+    public static final String CHANNEL_MEDIA_STOP = "mediaStop";
+    public static final String CHANNEL_APP_LAUNCHER = "appLauncher";
+    public static final String CHANNEL_MEDIA_PLAYER = "mediaPlayer";
+    public static final String PROPERTY_IP_ADDRESS = "IP";
```
well the ones above are channel IDs and this one is a thing property. the only concern I have is that this would break the configuration for current users.

martinvw replied:
Given its a property I don't think it will break, maybe you can quickly test it locally?

sprehn replied:
well it renders existing things unusable. but removing and rediscovering the device fixes it. so I'll change it to lowercase

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,237 @@
+            if (device.isConnected()) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connected");
+            } else {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connecting ...");
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java:
```diff
@@ -0,0 +1,237 @@
+
+    @Override
+    public void onConnectionFailed(ConnectableDevice device, ServiceCommandError error) {
+        logger.debug("Connection failed: {} - error: {}", device, error.getMessage());
```
Sure

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/ToastControlToast.java:
```diff
@@ -0,0 +1,60 @@
+public class ToastControlToast extends BaseChannelHandler<Void> {
+    private Logger logger = LoggerFactory.getLogger(ToastControlToast.class);
+
+    private ToastControl getControl(final ConnectableDevice device) {
```
Done

sprehn commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/ToastControlToast.java:
```diff
@@ -0,0 +1,60 @@
+            final String value = command.toString();
+            final ToastControl control = getControl(d);
+            try {
+                BufferedImage bi = ImageIO.read(getClass().getResource("/openhab-logo-square.png"));
```
Done. and sure, If there is a smarter way to fetch this... happy to amend

martinvw commented:
That is fast, thanks!

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+		<label>Command</label>
+		<description>Used to send ON commands to trigger an action. Or send ON and OFF commands to toggle state. Similar to buttons on a remote control.</description>
+	</channel-type>
+	<channel-type id="stringType">
```
remove this type and introduce dedicated ones for the functions

sprehn replied:
done

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+				<description>Current Mute Setting</description>
+			</channel>
+			<channel id="volume" typeId="volumeChannelType" />
+			<channel id="channel" typeId="stringType">
```
if it is a number, couldn't this be a Number channel instead? Then you could get rid off channelUp and channelDown as the number could be easily increased or decreased (e.g. through a setpoint widget).

sprehn replied:
will do. It is a string in connectsdk. on webos it is a parsable to a number.

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+				<description>Send a message onto the TV screen.</description>
+			</channel>
+			<channel id="mediaPlayer" typeId="mediaPlayerType" />
+			<channel id="mediaStop" typeId="commandType">
```
Can't this be removed as mediaPlayer already supports a "stop" command?

sprehn replied:
I see a PlayPauseType wich is mapped to play and pause on webos, but no equivalent stop type.

cweitkamp replied:
@kaikreuzer I agree with @sprehn . We use a separate channel for `stop` in the Kodi binding as well. I am not aware of the Player item to be capable of accepting something like a "stop" command.

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+			</channel>
+		</channels>
+
+		<!-- property as IP must not be changed -->
```
IP is clearly a required configuration parameter, so it must not be a property.

sprehn replied:
What is the difference between a property and a config parameter? 
I would think that users can modify the parameter.

Hilbrand replied:
A property is used for static meta data that can be set in the thing xml or when a thing is created  when discovered. It can not be changed later (It can but is not persisted if I read it correctly in the documentation) For example a serial number. A config parameter is for device specific settings, that could change or are user specific, like polling times or something.  Users can modify this parameter. An ip address is something that could change I presume so it's logical to have is as a config parameter, otherwise when it would change the user needs to re-add the binding.

kaikreuzer replied:
To even clarify a bit further: Configuration parameters configure the HANDLER, not the remote device. So the configuration needs all information that is required by the handler code to access the device. The IP address is very clearly such a piece of information. (see also https://github.com/eclipse/smarthome/issues/3484 if you think this is weird.)

sprehn commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,83 @@
+				<label>Power</label>
+				<description>Via this binding TV can only be powered off, not on.</description>
+			</channel>
+			<channel id="mute" typeId="commandType">
```
yes, that is the case already

martinvw commented:
LGTM, @kaikreuzer can you do the final honors.

sprehn commented:
Hi,

@kaikreuzer is there anything else expected from my end to get this merged?

I followed the recommendation to develop the next features in feature branches. It adds key controls (OK, Home, Up, Down, Left, Right) to the Action.
Shall I wait for this to merge before requesting the changes to be added in a new PR?

Kind Regards
Sebastian

sprehn commented:
Oh, and I may have just found away to make this binding completely independent from device IP addresses. I suggest to wait with the merge and I will add this change to this PR.

sprehn commented:
Here we go. I still wait for recommendation how to handle new features, such as key controls (OK, Home, Up, Down, Left, Right) to the Action.

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,223 @@
+ * This class provides the bridges between openhab thing discovery and connect sdk device discovery.
+ *
+ * @author Sebastian Prehn
+ * @since 1.8.0
```
remove the since

sprehn replied:
Sure

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSDiscovery.java:
```diff
@@ -0,0 +1,223 @@
+     * @param localIP optional configuration string
+     * @return local ip or <code>null</code> if detection was not possible.
+     */
+    private InetAddress findLocalInetAddresses(String localIP) {
```
You should not implement this yourself, but rather use the`org.eclipse.smarthome.core.net.NetworkAddressService` instead. This system service is already configurable and the user can set his local IP, which can then be used by ALL bindings, so there is no need for the binding to have such a configuration parameter at all.

sprehn replied:
Perfect.

sprehn replied:
I realize this is not yet available in my 2.1 installation.
But it works on 2.2.

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java.bak:
```diff
@@ -0,0 +1,81 @@
+package org.openhab.binding.lgwebos.internal.discovery;
```
I guess this file should be deleted?

sprehn replied:
Oh, this should have never been pushed. Will be deleted.

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -173,6 +173,11 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.lgtvserial/${project.version}</bundle>
     </feature>
 
+    <feature name="openhab-binding-lgwebos" description="LG webOS Binding" version="${project.version}">
```
Note that the Apache HTTP client libs are no longer part of a default installation. As your binding requires them, you need to explicitly add them here (like it was done [here](https://github.com/openhab/openhab2-addons/commit/b9d15820cf9b71b9ab46693ac21d6e6d7958758d)).

sprehn replied:
Done

kaikreuzer commented on addons/binding/org.openhab.binding.lgwebos/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,17 @@
+	<author>Sebastian Prehn</author>
+
+
+	<config-description>
```
This configuration should not be necessary, see my comment below on how to retrieve the local IP.

sprehn replied:
Done

sprehn commented:
Or, will remove the action from this PR. It would be interesting to see how we can avoid having to ship connect SDK and it dependencies a second time. Ideally the second action/rule module should then require this binding as a dependency. All it needs is a reference to LGWebOSDiscovery.

kaikreuzer commented:
>  Ideally the second action/rule module should then require this binding as a dependency.

Yes, this is probably the easiest solution. Alternatively, one could create a third bundle that can deal with the communication, but in this case here, it makes sense that the action directly requires a Thing (or rather its handler) in order to function.

sprehn commented:
@kaikreuzer, anything else expected from my end?

kaikreuzer commented:
No, this should be it - many thanks for your great work!

sprehn commented:
Thank you. 

@kaikreuzer, I just noticed reviewing the merge that the action class was not removed. I realize that I had deleted it locally, but obviously missed to commit the deletion. The other changes in Readme etc. went in as planned.
I prepared the file delete now in my fork. Should I open an Issue and raise a new pull request?

Also, where can I read about the new rule engine, so that I can start developing the module?
Only found this: http://docs.openhab.org/configuration/rules-ng.html

martinvw commented:
@sprehn no new issue is needed, you can just create a pr. the other question is for @kaikreuzer

sprehn commented:
thx, done: #2893

kaikreuzer commented:
@sprehn Docs are here: https://www.eclipse.org/smarthome/documentation/features/rules.html
It isn't much, but at least a start. Better to also look at existing implementations in the core, e.g. https://github.com/eclipse/smarthome/tree/master/bundles/automation/org.eclipse.smarthome.automation.module.media.
Note that we do not have any example for a real add-on yet. So some things will have to be worked out (like e.g. providing a skeleton archetype as we have for bindings, proper meta-data and documentation requirements, etc.).

sprehn commented:
Reading through this.. I guess I have to implement 
org.eclipse.smarthome.automation.handler.ModuleHandlerFactory and
org.eclipse.smarthome.automation.handler.ActionHandler

I am just wondering. Will we be able to use those actions also in traditional rule scripts? I personally rely on timers and some other features of the Xtend system.

kaikreuzer commented:
>  I guess I have to implement ...

Yes, sounds about right.

> Will we be able to use those actions also in traditional rule scripts?

Not yet, but the plan definitely is to merge the old and the new rule engine and make all "old" actions available in the new rule engine, while also giving access to new actions from within the Xbase scripted rules. I guess we will have to work out what kind of meta-data should be added for this, as atm it isn't straight-forward to get a method signature out of a (new) action type definition.

sprehn commented:
Started work on the action for the new rule engine. Created PR #2926 to facilitate the discussion.
